DAOS-7137 debug: add ABT ult id and xstream id to log

wangdi1 opened the pull request:
Add ABT ult id and xstream id to daos log. So daos log become

xxxx DAOS[process_id/xs_id/ABT thread ID] xxxxx

xs_id is -1 before tls is initialized, and thread id on client side,
otherwise it will be the xstream ID.

ABT thread ID is the ABT ult ID, 0 otherwise.

Signed-off-by: Di Wang <di.wang@intel.com>

wangdi1 commented:
I checked cart_logparser, and it seems we do not need change at all, since it does not try to ult/tid yet.

daosbuild1 commented on src/include/gurt/dlog.h:
```diff
@@ -371,11 +377,12 @@ void d_log_sync_mask(void);
  */
 int d_log_open(char *tag, int maxfac_hint, int default_mask,
-	       int stderr_mask, char *logfile, int flags);
+	       int stderr_mask, char *logfile, int flags, d_log_tid_cb_t tid_cb);
```
(style)  line over 80 characters

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-5255/1/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5255/1/execution/node/73/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

ashleypittman commented on src/cart/crt_group.c:
```diff
@@ -1179,33 +1179,30 @@ crt_group_lookup(crt_group_id_t grp_id)
 	return (grp_priv == NULL) ? NULL : &grp_priv->gp_pub;
 }
 
+/* NB: no D_DEBUG or D_ERROR in this function, since this might
+ * be executed inside the D_DEBUG(), see d_vlog().
+ */
```
You probably can't use D_ASSERT() or D_GOTO either then.

ashleypittman replied:
We're likely to either deprecate D_GOTO or land this patch https://github.com/daos-stack/daos/pull/3907 so it would be better not to use it in critical sections of the code.

ashleypittman commented:
How does this work with the log parser?

daosbuild1 commented:
Test stage Functional on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5255/2/execution/node/1239/log

daosbuild1 commented:
Test stage Functional on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5255/2/execution/node/1315/log

wangdi1 commented:
> How does this work with the log parser?

Hmm, you mean cart_logparser?   I saw it only retrieve the pid from [], then skip [], the pid will convert to rank by this patch.  Or I mss sth?

        fields = line.split()
        # Work out the end of the fixed-width portion, and the beginning of the
        # message.  The hostname and pid fields are both variable width
        idx = 29 + len(fields[1]) + len(fields[2])
        pidtid = fields[2][5:-1]
        pid = pidtid.split("/")
        self.pid = int(pid[0])
        self._preamble = line[:idx]
        self.fac = fields[3]

ashleypittman commented:
> > How does this work with the log parser?
> 
> Hmm, you mean cart_logparser? I saw it only retrieve the pid from [], then skip [], the pid will convert to rank by this patch. Or I mss sth?
> 
> ```
>     fields = line.split()
>     # Work out the end of the fixed-width portion, and the beginning of the
>     # message.  The hostname and pid fields are both variable width
>     idx = 29 + len(fields[1]) + len(fields[2])
>     pidtid = fields[2][5:-1]
>     pid = pidtid.split("/")
>     self.pid = int(pid[0])
>     self._preamble = line[:idx]
>     self.fac = fields[3]
> ```

So it just checks the bit before the first / then?  It should be unaffected by this.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5255/3/execution/node/1147/log

daosbuild1 commented:
Test stage Functional on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5255/3/execution/node/1285/log

daosbuild1 commented:
Test stage Functional on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5255/3/execution/node/1295/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Test CentOS 7 RPMs completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5255/4/execution/node/1147/log

daosbuild1 commented:
Test stage Functional on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5255/4/execution/node/1285/log

daosbuild1 commented:
Test stage Functional on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5255/4/execution/node/1295/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5255/5/execution/node/1264/log

daosbuild1 commented:
Test stage Functional on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5255/5/execution/node/1294/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

wangdi1 commented:
fix the CI test failure in cart_logparser.  since pid is being changed to rank, it can be 0 now. Please check. Thanks.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

jolivier23 commented on src/tests/ftest/cart/util/cart_logparse.py:
```diff
@@ -604,7 +604,7 @@ def new_iter(self,
 
         if stateful:
-            if not pid:
+            if pid is None:
```
Was it failing before due to getting back a 0?

wangdi1 replied:
yes, it does. since pid is being changed to rank with the patch.

wangdi1 commented:
@ashleypittman  Could you please review?  The log is very helpful for me to investigate ci failure. Thanks.

ashleypittman commented:
> @ashleypittman Could you please review? The log is very helpful for me to investigate ci failure. Thanks.

Just looking now.  CI shows 8 warnings fixed which is unlikely, and 117 new server memory leaks so the log parsing isn't picking this change up correctly yet.

wangdi1 commented:
> > @ashleypittman Could you please review? The log is very helpful for me to investigate ci failure. Thanks.
> 
> Just looking now. CI shows 8 warnings fixed which is unlikely, and 117 new server memory leaks so the log parsing isn't picking this change up correctly yet.

Oh, this is because process id is changed to  rank. So  before the cart is initialized, it uses the -1  until the rank is initialized by system leader.

So the log before this will be tracked by a -1 rank, instead of real rank, so you can see the "false" memleak here.

ashleypittman commented:
> > > @ashleypittman Could you please review? The log is very helpful for me to investigate ci failure. Thanks.
> > 
> > 
> > Just looking now. CI shows 8 warnings fixed which is unlikely, and 117 new server memory leaks so the log parsing isn't picking this change up correctly yet.
> 
> Oh, this is because process id is changed to rank. So before the cart is initialized, it uses the -1 until the rank is initialized by system leader.
> 
> So the log before this will be tracked by a -1 rank, instead of real rank, so you can see the "false" memleak here.

Yes, I've just came to the same conclusion.  As there are two ranks in the file (-1 and 0 in the case of NLT) then it assumes there are two processes and it parses them separately, the -1 process will appear to leak a lot of memory and I'd expect the 0 process to report some invalid frees but it doesn't appear to.

wangdi1 commented:
> > > > @ashleypittman Could you please review? The log is very helpful for me to investigate ci failure. Thanks.
> > > 
> > > 
> > > Just looking now. CI shows 8 warnings fixed which is unlikely, and 117 new server memory leaks so the log parsing isn't picking this change up correctly yet.
> > 
> > 
> > Oh, this is because process id is changed to rank. So before the cart is initialized, it uses the -1 until the rank is initialized by system leader.
> > So the log before this will be tracked by a -1 rank, instead of real rank, so you can see the "false" memleak here.
> 
> Yes, I've just came to the same conclusion. As there are two ranks in the file (-1 and 0 in the case of NLT) then it assumes there are two processes and it parses them separately, the -1 process will appear to leak a lot of memory and I'd expect the 0 process to report some invalid frees but it doesn't appear to.

Hmm, I am not sure about this, but I do debug the log parsing a bit. I can see it did  pick up the rank as the key(0, -1) correctly.  I am not sure why there is no invalid free here.  Are you sure this log parse can detect invalid free?

ashleypittman commented:
> Hmm, I am not sure about this, but I do debug the log parsing a bit. I can see it did pick up the rank as the key(0, -1) correctly. I am not sure why there is no invalid free here. Are you sure this log parse can detect invalid free?

See my update to the previous comment.  The log parser will pick up both invalid free and double-free.

I can update NLT to move the loop from log_parse into NLT itself but that'll just improve the reporting (to the extent that this PR would fail) but it won't fix the issue.

The code in question is here, it would be a case of moving the loop over "pids" from logtest to NLT itself, or catching the error, completing the loop and re-raising the error in the inner loop.
https://github.com/daos-stack/daos/blob/master/utils/node_local_test.py#L1562-L1571
https://github.com/daos-stack/daos/blob/master/src/tests/ftest/cart/util/cart_logtest.py#L289-L305

ashleypittman reviewed (changes requested):
See previous comments, this has issues with log parsing that are now showing up in CI results.

wangdi1 commented:
> See previous comments, this has issues with log parsing that are now showing up in CI results.
> 
> I think this is nearly there, I've just given a -1 to prevent accidental landing given it's green but with known issues.

Hmm, maybe I can make -1 match any rank it is assigned later, so it can track all allocation and free.  But I do not understand why the invalid free is not be shown up with the current change.

ashleypittman commented:
> > See previous comments, this has issues with log parsing that are now showing up in CI results.
> > I think this is nearly there, I've just given a -1 to prevent accidental landing given it's green but with known issues.
> 
> Hmm, maybe I can make -1 match any rank it is assigned later, so it can track all allocation and free. But I do not understand why the invalid free is not be shown up with the current change.

after errors in the first rank it stops processing further ranks.

wangdi1 commented:
> > > See previous comments, this has issues with log parsing that are now showing up in CI results.
> > > I think this is nearly there, I've just given a -1 to prevent accidental landing given it's green but with known issues.
> > 
> > 
> > Hmm, maybe I can make -1 match any rank it is assigned later, so it can track all allocation and free. But I do not understand why the invalid free is not be shown up with the current change.
> 
> after errors in the first rank it stops processing further ranks.

ah, I see. thanks.

wangdi1 commented:
@ashleypittman  I fixed the log_parser to parse all lines with single rank, please check.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Build RPM on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5255/8/execution/node/284/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5255/8/execution/node/334/log

daosbuild1 commented:
Test stage Build RPM on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5255/8/execution/node/347/log

ashleypittman commented on src/tests/ftest/cart/util/cart_logparse.py:
```diff
@@ -535,7 +535,12 @@ def _load_data(self):
                     pids[l_pid]['line_count'] += 1
                 else:
-                    pids[l_pid] = {'line_count': 1,
+                    if l_pid != int(-1) and -1 in pids:
```
Is this the bit that's new?  If so can you comment on what it's doing please?

If I remember this is the part that creates some metadata about pids and their first character and line count in the file so it can be parsed from disk, but there are two other implementations of this, one for small files where the entire file is loaded into memory, and one for bzip files where it's walked iteratively but you can't seek() so it has to start at the beginning each time.

wangdi1 replied:
Inside the log, the rank(pid) is -1 before the node is assigned to the real rank. So this is for checking if the rank is assigned (!= -1) and there -1 in pid(previous log),  then let's merge them together.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

ashleypittman reviewed (dismissed):
The python changes look correct now but I'm having trouble running it locally (daos engine is exiting on startup) so I'll check the CI logs once it's built.

daosbuild1 commented:
Test stage Functional on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5255/9/execution/node/1152/log

daosbuild1 commented:
Test stage Functional on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5255/9/execution/node/1162/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional on CentOS 7 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5255/11/execution/node/1117/log

daosbuild1 commented:
Test stage Functional on Leap 15 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-5255/11/execution/node/1163/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.
